### ISSUE TITLE
Does Logstash log the misused Aggregate action ? (update action called on not created map)

### ISSUE THREAD

**Issue by alkuzad** [logstash-plugins/logstash-filter-aggregate#19]
I was wondering if currently there is an option to see if aggregate was called with map_action update on event that does not have map created yet (--debug option for logstash ?)

Background:

>  I had problem when some of my data in Elasticsearch had missing fields (maybe 5%). I've lowered      number of concurrent workers pushing data to Redis and situation got better but still some of events were incomplete. Then I found that I have my aggregate setup wrong and some map creations were not created in scenario one event was sent too fast (push workers are not guaranteed to send events to redis in same order) and called with update on create action. After fixup all events are good.

So - is there an option to check it in logs ? I see that on misaction filter simply returns:

``` ruby
   aggregate_maps_element = @@aggregate_maps[task_id]
      if (aggregate_maps_element.nil?)
        return if @map_action == "update"
```

But does it have impact on logging ? So I can see like "no filter action taken" or something ?

Could we potentially add something like "tag_on_misaction" which will add custom tag on such events ? Like `tag_on_misaction => "_aggregate_update_on_mapless_event"`

**Comment by fbaligand** [comment]
Hi @alkuzad,

Several things :
- firstable, if your filter workers are not set to 1 (which is not anymore the default value since logstash 2.0), you have no guaranty on events order, and aggregate plugin would not like it. If events doesn't arrive in the good order, you will have exactly the problem you quote.
- then, presently, aggregate plugin does not log "misactions". But I really think this would be a nice enhancement.
- I can log at info level, with event logged when :
 - map is not created and event arrived while map_action=update
 - map is already created and event arrived while map_action=create
 - timeout occurs on a map (because no "end_of_task" happened before)
- that said, if I log these use cases, do you really need an extra tag ?

**Comment by alkuzad** [comment]
Hi @fbaligand 

I'm using Logstash 1.5 for now so the problem with filter workers is not my issue. I also use external script that shares informations with Logstash via redis and ensures only after all events from one batch were processed by Logstash it will send the aggregated event to ES. The only problem was that LS sent "ok job is done" message and event was parsed but content was not added to aggregate map.

It wasn't my intention to add logging to Aggregate, only to figure out if I can debug it with current Logstash/Aggregate functionality. If it's not possible then I think tagging is better as it's lot of easier to process them. Also using info level log feels like inception - using logstash for parsing logs and then another one for parsing logs from master logstash if any problems occured ?

**Comment by fbaligand** [comment]
OK.

If you need to quickly debug, you can make debug aggregate configurations which add a tag when you are not in opposite expected state.

For example, if you expect "map_action=>update", you can try the following config :

```
aggregate {
    task_id => "..."
    map_action=> "update"
    code=> "do the right code"
}
aggregate {
    task_id => "..."
    map_action=> "create"
    code=> "do whatever code with this bad state..."
    add_tag => ["map_not_created"]
}
```

**Comment by alkuzad** [comment]
@fbaligand thanks for the tip :+1: Unfortunately I have over 100 aggregates now (almost each in own conditional) so adding next 100 would be quite heavy :)

**Comment by fbaligand** [comment]
You can create a new logstash config file, named "debug.conf", where you copy/paste all your aggregate conf and then make some generic search/replace. Maybe it could be done quickly. 

Then you add this conf at logstash startup and you're done. 
At production startup, you remove this "debug config file" and that's nice. 

Fabien

**Comment by fbaligand** [comment]
Hi @alkuzad 

1 year after you opened the issue (^^), I had a lightning idea for you problem.
Currently, you can't add tag if the code is not executed. But on the contrary, if the code is executed successfully, you can add/remove a tag or add/remove a field, using standard Logstash filter options :
`add_tag, add_field, remove_tag, remove_field`. Actually, these options are applied only if code is executed successfully.

Thus, in the next release (2.5.1), I will add debug logging, which you can activate only for aggregate plugin using log4j2 in Logstash 5.

**Comment by fbaligand** [comment]
With debug logs added in v2.5.1, this issue is now fixed.

**Comment by alkuzad** [comment]
Thanks @fbaligand :) It's nice but I'm not using aggregate any more, and I'm changing infrastructure logging to work without that need. Main problem is that aggregate by design can not scale and we started to see that one LS only, even if powerful, is big bottleneck. And on top of that Redis also can not scale for list-based database (only key-value storage is well scalable) and I was using it as backup to store data before we fixed LS again.

**Comment by fbaligand** [comment]
Hi @alkuzad 
For my company needs, I use Logstash and aggregate plugin.
As you, I had scale issues.
And this is the different ways I addressed theses issues :
- I set one Logstash instance per source host
- I set a centralized cluster Logstash cluster (multiple instances on one host), each instance listening on a different port. I set one Beats per source host that transfers data to one specific Logstash cluster instance
- I set a centralized cluster Logstash cluster, but with a kafka cluster in front. I set one Beats per source host that transfers data to the kafka cluster on topic "${HOSTNAME}". On each Logstash cluster instance, I listen on a kafka topics whitelist.
